iroha/src/network_topology.rs

```diff
@@ -0,0 +1,197 @@
+use iroha_crypto::Hash;
+use iroha_data_model::prelude::*;
+use rand::{rngs::StdRng, seq::SliceRandom, SeedableRng};
+use std::{
+    collections::BTreeSet,
+    fmt::Debug,
+};
+
+
+/// Uninitialized `NetworkTopology`, use only for construction.
+/// Call `init` to get `InitializedNetworkTopology` and access all other methods.
+#[derive(Debug)]
+pub struct NetworkTopology {
+    peers: BTreeSet<PeerId>,
+    max_faults: usize,
+    block_hash: Option<Hash>,
+}
+
+impl NetworkTopology {
+    /// Constructs a new `NetworkTopology` instance.
+    pub fn new(
+        peers: &BTreeSet<PeerId>,
+        block_hash: Option<Hash>,
+        max_faults: usize,
+    ) -> NetworkTopology {
+        NetworkTopology {
+            peers: peers.clone(),
+            max_faults,
+            block_hash,
+        }
+    }
+
+    /// Initializes network topology.
+    pub fn init(self) -> Result<InitializedNetworkTopology, String> {
+        let min_peers = 3 * self.max_faults + 1;
+        if self.peers.len() >= min_peers {
+            let mut topology = InitializedNetworkTopology {
+                sorted_peers: self.peers.into_iter().collect(),
+                max_faults: self.max_faults,
+            };
+            topology.sort_peers(self.block_hash);
+            Ok(topology)
+        } else {
+            Err(format!("Not enough peers to be Byzantine fault tolerant. Expected a least {} peers, got {}", 3 * self.max_faults + 1, self.peers.len()))
+        }
+    }
+}
+
+/// Represents a topology of peers, defining a `role` for each peer based on the previous block hash.
+#[derive(Debug, Clone)]
+pub struct InitializedNetworkTopology {
+    /// Current order of peers. The roles of peers are defined based on this order.
+    pub sorted_peers: Vec<PeerId>,
+    max_faults: usize,
+}
```

Comment:
I suggest to leave `NetworkTopology` in the `sumeragi` module as the whole topology ordering and role distribution are very consensus specific in this case.

We have an issue to support many different consensus algos and therefore to abstract it away with some common interface, but it will require a lot of work and is of low priority now.